cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
     }
 }
 
+/// Get the number of general-purpose and floating-point registers required to
+/// hold the given ABI parameters.
+fn num_registers_required<'a>(
+    word_bit_size: u8,
+    call_conv: CallConv,
+    shared_flags: &shared_settings::Flags,
+    isa_flags: &isa_settings::Flags,
+    params: impl IntoIterator<Item = &'a AbiParam>,
+) -> (usize, usize) {
+    let gprs = &[RU::rax; 128];
+    let mut assigner = Args::new(
+        word_bit_size,
+        gprs,
+        std::usize::MAX,
+        call_conv,
+        shared_flags,
+        isa_flags,
+    );
+
+    let mut gprs_required = 0;
+    let mut fprs_required = 0;
+
+    for p in params {
+        // We're going to mutate the type as it gets converted, so make our own
+        // copy that isn't visible to the outside world.
+        let mut p = p.clone();
+
+        let mut split_factor = 1;
+
+        'converting_param: loop {
+            match assigner.assign(&p) {
+                ArgAction::Convert(ValueConversion::IntSplit) => {
+                    split_factor *= 2;
+                    p.value_type = p.value_type.half_width().unwrap();
+                }
+                ArgAction::Convert(ValueConversion::VectorSplit) => {
+                    split_factor *= 2;
+                    p.value_type = p.value_type.half_vector().unwrap();
+                }
+                ArgAction::Assign(ArgumentLoc::Reg(_))
+                | ArgAction::Convert(ValueConversion::IntBits)
```

Comment:
That would be nice, excpet `legalize_args` mutates the parameters, and we might need to undo those mutations if we decide to use a struct return pointer instead of registers. For example, `legalize_args` might split values into smaller pieces with less alignment requirements, that we then need to reconstruct and ensure are aligned if we store it through the struct return pointer.

The easiest way to do that would be to keep a clone of the vec of original parameters that we can restore if necessary, but that clone has a potential perf implication.

Would you prefer to do that clone rather than count the registers required?

Sort of an aside: assigning parameters and returns to registers would really benefit from some re-working to make backtracking easier, as sys v (and I'm sure other calling ocnventions too) need that in various scenarios... :-/